### PR TITLE
Support sub-representations

### PR DESCRIPTION
**Pull request by RackerWilliams**
Allow a representation within a representation. This is a sub representation.

````xml
<wadl:representation mediaType="application/xml" >
     <rax:representation mediaType="application/json" path="/path/to/json" name="embeddedjson">
        <wadl:param style="plain" path="$_?name='foo'" /> 
      </rax:representation> 
</wadl:representation>
````

In the example above json is in XML at /path/to/json.

The following rules will apply:

1. ```rax:represantation``` can only live inside a ```wadl:representation``` or another ```rax:representation``` -- notice ```rax:``` instead of ```wadl:``` this is because wadl doesn't allow sub-representation -- it's an extension.
2. works exactly like wadl:representation  all features of representation are supported:
     1. plain parameters
     2. XSD / JSON Schema check
     3. Assertions
     4. Catpture header
3. Changes made to a subrepresntation are not carried over to the origin service...this will be handled in another PR.
4.  Only XML and JSON representations are supported...other mediaTypes are ignored.

### CODE REVIEW

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala [hunk 1]
```diff
@@ -0,0 +1,1387 @@
...
+      validator.validate(request("PUT", "/a", "application/xml", goodRepXML1, true,
+        Map[String,List[String]]("X-ROLES"->List("admin"))),response, chain)
+    }
+
+    test(s"A PUT on /a should succeed with good XML 2 $wadlDesc with $configDesc") {
+      validator.validate(request("PUT", "/a", "application/xml", goodRepXML2, true,
+        Map[String,List[String]]("X-ROLES"->List("allowPUT"))),response, chain)
+    }
+
+    test(s"A PUT on /a should succeed with good XML 3 if it is specifed in the X-XML-USER header $wadlDesc with $configDesc") {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: Typo: specified

Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: 40+ places.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala [hunk 2]
```diff
@@ -0,0 +1,1387 @@
...
+    test(s"A PUT on /a/b should fail if bad XML is specifed in the X-XML-USER header if rax:rep is enabled (yaml) $wadlDesc with $configDesc") {
+      assertResultFailed(validator.validate(request("PUT", "/a/b", "text/yaml", goodRepYAML1, false,
+        Map[String,List[String]]("X-ROLES"->List("wooga"),
+          "X-XML-USER"->List("!This is not XML!"),
+          "X-XML-USER2"->List("!This is not XML!"),
+          "X-JSON-USER"->List("!This is not JSON!"))),response, chain),
+        400, List("Bad Content"))
+    }
+
+    test(s"A PUT on /a/b should fail with bad embeded XML if rax:rep is enabled $wadlDesc with $configDesc") {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: Typo: embedded

Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: 10+ places

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala [hunk 3]
```diff
@@ -0,0 +1,1387 @@
...
+      </xml>
+    </other_xml>
+
+
+
+  //
+  //  Assertions!
+  //
+  def happyPathAssertions(validator : Validator, wadlDesc : String, configDesc : String) {
+    test(s"A PUT on /a should succeed with good XML 1 $wadlDesc with $configDesc") {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: All of these tests that are similar could be rolled up with a `Seq().foreach` like these:
 - https://github.com/rackerlabs/api-checker/blob/api-checker-2.5.1/core/src/test/scala/com/rackspace/com/papi/components/checker/handler/ApiCoverageHandlerTest.scala#L76-L89

Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: That roll up doesn't save a lot of code in these cases, but it does provide a grouping for similar tests.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala [hunk 4]
```diff
@@ -0,0 +1,1402 @@
...
+            <param name="test6" style="plain"
+                   path="tst:user2/@ln" required="true"
+                   rax:message="Need a last name"
+                   rax:code="403"/>
+        </rax:representation>
+    </resources>
+  </application>
+
+
+  feature ("The WADLCheckerBuilder can correctly transforma a WADL into checker format") {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Typo: transforma

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala [hunk 5]
```diff
@@ -0,0 +1,1402 @@
...
+      assert(checker, Start, URL("a"), Method("POST"), ReqType("(application/json)(;.*)?"), WellJSON, RaxPushXML("xmlUser","req:header('X-XML-USER')"),
+        XPath("/tst:user"), XPath("tst:user/@firstName"), XPath("tst:user/@lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("POST"), ReqType("(text/yaml)(;.*)?"), RaxPushXML("xmlUser","req:header('X-XML-USER')"),
+        XPath("/tst:user"), XPath("tst:user/@firstName"), XPath("tst:user/@lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), URL("b"), Method("PUT"), ReqType("(application/xml)(;.*)?"), WellXML, XPath("/tst:other_xml"), Accept)
+      assert(checker, Start, URL("a"), URL("b"), Method("PUT"), ReqType("(application/json)(;.*)?"), WellJSON, Accept)
+      assert(checker, Start, URL("a"), URL("b"), Method("PUT"), ReqType("(text/yaml)(;.*)?"), Accept)
+    }
+
+    scenario ("The WADL contains rax:rep at the representation at multiple levels, at method with multiple reps, and at the resource level, applyChildernFalse  (rax:representation enabled)") {
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Typo: Children

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala [hunk 6]
```diff
@@ -0,0 +1,1402 @@
...
+        XPath("/tst2:user"), RaxPushJSON("/tst2:user/tst2:json"), JsonXPath("$_?lastName"), RaxPopRep, RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("PUT"), ReqType("(application/json)(;.*)?"), WellJSON, JsonXPath("$_?firstName"), RaxPushXML("xmlUser","req:header('X-XML-USER')"),
+        XPath("/tst:user"), XPath("tst:user/@firstName"), XPath("tst:user/@lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("POST"), ReqType("(application/xml)(;.*)?"), WellXML, XPath("/tst:some_xml"), RaxPushJSON("jsonUser","req:header('X-JSON-USER')"),
+        JsonXPath("$_?firstName"), JsonXPath("$_?lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("POST"), ReqType("(application/json)(;.*)?"), WellJSON, RaxPushJSON("jsonUser","req:header('X-JSON-USER')"),
+        JsonXPath("$_?firstName"), JsonXPath("$_?lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("POST"), ReqType("(text/yaml)(;.*)?"), RaxPushJSON("jsonUser","req:header('X-JSON-USER')"),
+        JsonXPath("$_?firstName"), JsonXPath("$_?lastName"), RaxPopRep, Accept)
+      assert(checker, Start, URL("a"), Method("POST"
```
Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Typo: Children

Comment (wdschei) [inline core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Typo: Children

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala (hunk 3 above)
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: I like it, I'll refactor these and follow that pattern in the future.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala [hunk 7]
```diff
@@ -7042,4 +7048,1132 @@ class StepSuite extends BaseStepSuite with MockitoSugar {
...
+  }
+
+  test("In a PushXML step, we should be able to push an XML representation from a request body") {
+    val nsContext = ImmutableNamespaceContext(Map[String,String]())
+    val pushXML = new PushXML("PushXML", "PushXML", "XML", "$body('xml')",nsContext,31,10,Array[Step]())
+    val req = request("PUT", "/a/b","application/json","""
+         {
+            "xml" : "<foo xmlns='http://rackspace.com/bar' />"
+         }
+    """, true)
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala]: No, whitespace shouldn't matter, just trying to get my code to look nice!  I'll use the pattern you suggest.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala [hunk 8]
```diff
@@ -7042,4 +7048,1132 @@ class StepSuite extends BaseStepSuite with MockitoSugar {
...
+    val origRep = req.parsedRepresentation
+    assert(origRep.isInstanceOf[ParsedJSON])
+    val pushContext = pushXML.checkStep(req, response, chain, StepContext())
+
+    //
+    //  Context should not be defined
+    //
+    assert(!pushContext.isDefined)
```
[the remark this answers is not in the document]
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala]: Ah okay, I can see how that makes things a bit cleaner.  I'll make the change.

**Comment by RackerWilliams** [comment]
Okay, I'll fix the typos and the changes above, expect a fix soonish.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala (hunk 3 above)
Comment (RackerWilliams) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: Actually, I think I'm going to skip this actual fix and save it for the next time I do this kind of test.  You're right that it doesn't save that much code, and I need to move on to the next PR.

**Comment by RackerWilliams** [comment]
@wdschei  Okay pushed all fixes except the one about refactoring `.foreach` -- see the comment above. Let me know if I missed something.

File: core/src/main/resources/xsl/checker-asserts.xsl [hunk 9]
```diff
@@ -224,35 +224,61 @@
...
         <!--
             Content dependecies:  XSD and XPATH require parent to be XSL, WELL_XML, XPATH
                                   JSON_SCHEMA and JSON_XPATH requires parent to be WELL_JSON
+
+                                  Both can have a parent of type
+                                  PUSH_REP, but it must be of the right type.
+
+                                  Both can have a parent of a POP_REP,
+                                  because we might be popping into the
+                                  correct representation.  TODO: beter
```
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/main/resources/xsl/checker-asserts.xsl]: :+1:

File: core/src/main/resources/xsl/checker-asserts.xsl [hunk 10]
```diff
@@ -224,35 +224,61 @@
...
+                                  checks on poping correctly!
```
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/main/resources/xsl/checker-asserts.xsl]: Not a big deal, but this same typo is also in the `TODO.org`.

File: core/src/main/resources/xsl/checker-asserts.xsl [hunk 11]
```diff
@@ -224,35 +224,61 @@
...
+                    </xsl:message>
+                </xsl:if>
+            </xsl:when>
         </xsl:choose>
     </xsl:template>
 
+    <xsl:template match="chk:step[@type='POP_REP']">
+        <xsl:if test="not(chk:ancestorTypes(.,('PUSH_XML_REP','PUSH_JSON_REP')))">
+            <xsl:message terminate="yes">
+                The POP_REP Step <xsl:value-of select="@id"/> must eventually be proceeded by a PUSH_REP.
```
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/main/resources/xsl/checker-asserts.xsl]: :+1:

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala (hunk 3 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: :+1:

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala (hunk 1 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: Not a big deal, but this same typo is in 3 other places in 2 different files.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala (hunk 2 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/ValidatorWADLRAXRepresentationSuite.scala]: Not a big deal, but this same typo is in 5 other places in 3 different files.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala]: :+1:

File: core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala]: :+1:

File: core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala [hunk 12]
```diff
@@ -7042,4 +7048,1132 @@ class StepSuite extends BaseStepSuite with MockitoSugar {
...
+
+    //
+    //  Orginial rep should stay in place
```
[the remark this answers is not in the document]
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/step/StepSuite.scala]: :+1:

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala (hunk 4 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Not a big deal, but this same typo is in 13 other places in 13 different files.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala (hunk 5 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Not a big deal, but this same typo is in 3 other places in 2 different files.

File: core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala (hunk 6 above)
Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: I'm not exactly where the original comment was supposed to be, but the following typos were near it:
 - Childred - 2 places in 1 file.
 - Chlidren - 17 places in 3 files.

Comment (wdschei) [reply core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerRaxRepStepSpec.scala]: Same as above.

**Comment by wdschei** [comment]
That makes sense about the `.foreach`. We should probably get a story in on our backlog to just go in and do some non-functional cleanup of a few things like this and a blanket typo fix.

**Pull request merged**